installer/lib/phx_new/web.ex

```diff
@@ -9,7 +9,7 @@ defmodule Phx.New.Web do
     {:config, "#{@pre}/config/config.exs",            :project, "config/config.exs"},
     {:config, "#{@pre}/config/dev.exs",               :project, "config/dev.exs"},
     {:config, "#{@pre}/config/prod.exs",              :project, "config/prod.exs"},
-    {:config, "#{@pre}/config/prod.secret.exs",       :project, "config/prod.secret.exs"},
+    {:config, "#{@pre}/config/runtime.exs",           :project, "config/runtime.exs"},
```

Comment:
I think this should be:

```suggestion
    {:prod_config, "#{@pre}/config/runtime.exs",           :project, "config/runtime.exs"},
```

And then in here:

https://github.com/phoenixframework/phoenix/blob/0048cd7a239982b7b4c6c20de780886b9027dee1/installer/lib/phx_new/generator.ex#L58

You will add a new clause that is exactly the same as `:config`, except it calls your new `prod_only_config_inject` function (great job on it btw). If you don't do this change, `web.ex` will inject contents outside of the `config_env() == :prod` block. :)

Reply:
@josevalim I've tried a couple of approaches now:

First, using the [current web config/runtime.exs](installer/templates/phx_umbrella/apps/app_name_web/config/runtime.exs), the contents of that file are appended after `import Config`. Any other prod-specific configs are injected using the `prod_only_config_inject` function. This results in the app_root/config/runtime.exs file that looks like what I would expect:
<details>
  <summary>current config/runtime.exs</summary>

```elixir
# In this file, we load production configuration and secrets
# from environment variables. You can also hardcode secrets,
# although such is generally not recommended and you have to
# remember to add this file to your .gitignore.
import Config

if config_env() == :prod do
  database_url =
    System.get_env("DATABASE_URL") ||
      raise """
      environment variable DATABASE_URL is missing.
      For example: ecto://USER:PASS@HOST/DATABASE
      """

  config :dev_app, DevApp.Repo,
    # ssl: true,
    url: database_url,
    pool_size: String.to_integer(System.get_env("POOL_SIZE") || "10")

  secret_key_base =
    System.get_env("SECRET_KEY_BASE") ||
      raise """
      environment variable SECRET_KEY_BASE is missing.
      You can generate one by calling: mix phx.gen.secret
      """

  config :dev_app_web, DevAppWeb.Endpoint,
    http: [
      # Enable IPv6 and bind on all interfaces.
      # Set it to  {0, 0, 0, 0, 0, 0, 0, 1} for local network only access.
      ip: {0, 0, 0, 0, 0, 0, 0, 0},
      port: String.to_integer(System.get_env("PORT") || "4000")
    ],
    secret_key_base: secret_key_base

  # ## Using releases
  #
  # If you are doing OTP releases, you need to instruct Phoenix
  # to start each relevant endpoint:
  #
  #     config :dev_app_web, DevAppWeb.Endpoint, server: true
  #
  # Then you can assemble a release by calling `mix release`.
  # See `mix help release` for more information.
end
```
</details>

This has the downside of being order dependent: The expectation is that the web config will be copied before the DB config (or any other prod-spcific config).

---

I also tried the approach you suggested (using `prod_only_config_inject` to inject the web config into app_root/config/runtime.exs, which contains `if config_env() == :prod do`). This approach was a little smarter about indentation the web config didn't have to have the environment check and didn't require indentation. However, I've had a hard time getting the formatting correct with this approach: `prod_only_config_inject` will insert a newline before and after the injected text. This leads to a config file which isn't formatted correctly (specifically, the newline before `end`):

<details>
  <summary>injected config/runtime.exs</summary>

```elixir
# In this file, we load production configuration and secrets
# from environment variables. You can also hardcode secrets,
# although such is generally not recommended and you have to
# remember to add this file to your .gitignore.
import Config

if config_env() == :prod do
  database_url =
    System.get_env("DATABASE_URL") ||
      raise """
      environment variable DATABASE_URL is missing.
      For example: ecto://USER:PASS@HOST/DATABASE
      """

  config :dev_app, DevApp.Repo,
    # ssl: true,
    url: database_url,
    pool_size: String.to_integer(System.get_env("POOL_SIZE") || "10")

  secret_key_base =
    System.get_env("SECRET_KEY_BASE") ||
      raise """
      environment variable SECRET_KEY_BASE is missing.
      You can generate one by calling: mix phx.gen.secret
      """

  config :dev_app_web, DevAppWeb.Endpoint,
    http: [
      # Enable IPv6 and bind on all interfaces.
      # Set it to  {0, 0, 0, 0, 0, 0, 0, 1} for local network only access.
      ip: {0, 0, 0, 0, 0, 0, 0, 0},
      port: String.to_integer(System.get_env("PORT") || "4000")
    ],
    secret_key_base: secret_key_base

  # ## Using releases
  #
  # If you are doing OTP releases, you need to instruct Phoenix
  # to start each relevant endpoint:
  #
  #     config :dev_app_web, DevAppWeb.Endpoint, server: true
  #
  # Then you can assemble a release by calling `mix release`.
  # See `mix help release` for more information.

end
```
</details>

I'm sure there's an ergonomic way around this, but I haven't hit it yet. This approach has the upside of being order-independent, but I'm not quite there with the implementation. I can push a commit that shows this work if you're interested in seeing it. I can do some commit cleanup once the direction is decided.

Reply:
I think what you can do for now is to go with the second approach and automatically amend the end. If you see this: "\n\nend\n", you can rewrite it to "\nend\n". I think this transformation is always safe. :)

Reply:
Well, there is another approach you can take that actually solves all of this: just inject the contents, without indentation nor anything, and then call `Code.format_string!`. :D It will indent and resolve everything for us. Plus, even better, it will guarantee it is formatted. We can also call it for the regular config_inject too!

Reply:
I haven't figured out how to plug in `Code.format_string!` yet. I've tried a few variations of:
```elixir
    case split_with_self(contents, "if config_env() == :prod do") do
      [left, middle, right] ->
        formatted_contents =
          Enum.join([left, middle, ?\n, indent(to_inject, 2), ?\n, right])
          |> Code.format_string!()
        File.write!(file, formatted_contents)
```

Each time I end up with exceptions like:
```
** (SyntaxError) nofile:32:1: unexpected reserved word: end
    (elixir 1.11.1) lib/code.ex:654: Code.format_string!/2
```

I've also tried formatting the file in place:
```elixir
      [left, middle, right] ->
        File.write!(file, [left, middle, ?\n, indent(to_inject, 2), ?\n, right])
        Code.format_file!(file)
```

The result, however, is the same as my non-formatted output (which fails the formatting check).

Any tips for using `Code.format_#{thing}`?

Reply:
If format is failing is because the file is indeed invalid. Can you please print the contents of the file before formatting?

Reply:
Also, you shouldn’t need indent anymore. :)

Reply:
This is the output of `IO.inspect` when I shim it into the pipeline:

```elixir
"# In this file, we load production configuration and secrets\n# from environment variables. You can also hardcode secrets,\n# although such is generally not recommended and you have to\n# remember to add this file to your .gitignore.\nimport Config\n\nif config_env() == :prod do10  secret_key_base =\n    System.get_env(\"SECRET_KEY_BASE\") ||\n      raise \"\"\"\n      environment variable SECRET_KEY_BASE is missing.\n      You can generate one by calling: mix phx.gen.secret\n      \"\"\"\n  \n  config :dev_app_web, DevAppWeb.Endpoint,\n    http: [\n      # Enable IPv6 and bind on all interfaces.\n      # Set it to  {0, 0, 0, 0, 0, 0, 0, 1} for local network only access.\n      ip: {0, 0, 0, 0, 0, 0, 0, 0},\n      port: String.to_integer(System.get_env(\"PORT\") || \"4000\")\n    ],\n    secret_key_base: secret_key_base\n  \n  # ## Using releases\n  #\n  # If you are doing OTP releases, you need to instruct Phoenix\n  # to start each relevant endpoint:\n  #\n  #     config :dev_app_web, DevAppWeb.Endpoint, server: true\n  #\n  # Then you can assemble a release by calling `mix release`.\n  # See `mix help release` for more information.10\nend\n"
```

Reply:
You can see a 10 after the “do”. That’s because Enum.join is converting ?\n to 10. Either use “IO.iodata_to_binary” instead of Enum.join or use “\n”. I think iodata_to_binary is better!

Reply:
That did the trick. Thank you!